### PR TITLE
Implement /unblacklistall command

### PR DESCRIPTION
**Pull request by panpawn**
This also introduces Punishments#roomUnblacklistAll to do the heavy lifting of this command.

### CODE REVIEW

**Comment by Zarel** [comment]
...who asked for this? I don't think this is a good idea... especially if used accidentally.

File: commands.js [hunk 1]
```diff
@@ -2149,6 +2149,25 @@ let commands = exports.commands = {
 	},
 	unblacklisthelp: ["/unblacklist [username] - Unblacklists the user from the room you are in. Requires: # & ~"],
 
+	unblacklistall: function (target, room, user) {
+		if (!this.can('editroom', null, room)) return false;
+
+		if (!target) {
+			user.lastCommand = '/unblacklistall';
+			this.errorReply("THIS WILL UNBLACKLIST ALL BLACKLISTED USERS.");
+			this.errorReply("To confirm, use: /unblacklistall confirm");
```
Comment (panpawn) [inline commands.js]: This kind of prevents the "if used accidentally" scenario, for the most part

**Comment by panpawn** [comment]
Also, I just think it might be helpful with rooms where the blacklists are rather large and if a room should want to remove them; same purpose for ``/deroomvoiceall``.

Also, I thought we were making roomban-type punishments closer to global ban punishments... There is ``'unbanall``, so I thought with that in mind, mirroring that functionality wouldn't be a horrible idea.

**Comment by bumbadadabum** [comment]
I intentionally didn't add this tbh

File: punishments.js [hunk 2]
```diff
@@ -726,6 +726,23 @@ Punishments.roomUnblacklist = function (room, userid) {
...
+
+	roombans.forEach((punishment, userid) => {
+		if (punishment[0] === 'BLACKLIST') {
+			Punishments.roomUnblacklist(room, userid);
+			success = true;
+		}
+	});
+	if (!success) return false;
+	Punishments.savePunishments();
+	return success;
```
Comment (Slayer95) [inline punishments.js]: This should probably return the list of previously blacklisted users, which should be logged.
With that, the command becomes revertible and safe.

Comment (panpawn) [inline punishments.js]: Good idea!

Comment (panpawn) [inline punishments.js]: Done

**Comment by panpawn** [comment]
Updated

File: commands.js [hunk 3]
```diff
@@ -2149,6 +2149,25 @@ let commands = exports.commands = {
...
+			return;
+		}
+		if (user.lastCommand !== '/unblacklistall' || target !== 'confirm') {
+			return this.parse('/help unblacklistall');
+		}
+
+		let unblacklisted = Punishments.roomUnblacklistAll(room);
+		if (!unblacklisted) return this.errorReply("No users are currently blacklisted in this room to unblacklist.");
+		this.addModCommand("All blacklists in this room have been lifted by " + user.name + ".  This has unblacklisted: " + unblacklisted.join(', '));
```
Comment (Slayer95) [inline commands.js]: Like this it will spam the room though.

Just log it with ``this.logEntry``

Comment (panpawn) [inline commands.js]: Done

**Comment by panpawn** [comment]
Updated

File: punishments.js (hunk 2 above)
Comment (Zarel) [inline punishments.js]: Well, it's not completely revertible, unless you log IPs/alts, and even then there's not a good way to add IPs/alts to a blacklist.

File: commands.js [hunk 4]
```diff
@@ -2149,6 +2149,27 @@ let commands = exports.commands = {
...
+		}
+		if (user.lastCommand !== '/unblacklistall' || target !== 'confirm') {
+			return this.parse('/help unblacklistall');
+		}
+
+		let unblacklisted = Punishments.roomUnblacklistAll(room);
+		if (!unblacklisted) return this.errorReply("No users are currently blacklisted in this room to unblacklist.");
+		let reply = `All blacklists in this room have been lifted by ${user.name}.`;
+		room.add(`|c|${user.getIdentity()}|/log ${reply}`).update();
+		this.logEntry(`${reply}  This has unblacklisted: ${unblacklisted.join(', ')}`);
```
Comment (panpawn) [inline commands.js]: Funny you mention that actually, I was taught to always put two spaces in-between sentences in school when I was learning how to type.  I'll remove this though, now that I'm remembering this isn't a standard for PS stuff.

**Comment by panpawn** [comment]
Updated

File: commands.js [hunk 5]
```diff
@@ -2149,6 +2149,27 @@ let commands = exports.commands = {
...
+		}
+		if (user.lastCommand !== '/unblacklistall' || target !== 'confirm') {
+			return this.parse('/help unblacklistall');
+		}
+
+		let unblacklisted = Punishments.roomUnblacklistAll(room);
+		if (!unblacklisted) return this.errorReply("No users are currently blacklisted in this room to unblacklist.");
+		let reply = `All blacklists in this room have been lifted by ${user.name}.`;
+		room.add(`|c|${user.getIdentity()}|/log ${reply}`).update();
+		this.logEntry(`${reply} This has unblacklisted: ${unblacklisted.join(', ')}`);
```
Comment (Zarel) [inline commands.js]: I think this is redundant with the `room.add` above. I think the `${reply}` part can be safely removed.

Comment (panpawn) [inline commands.js]: So, how are we mod-logging this then?

Comment (Zarel) [inline commands.js]: The code I recommend here is:

```
this.addModCommand(`All blacklists in this room have been lifted by ${user.name}.`);
this.logEntry(`Unblacklisted users: ${unblacklisted.join(', ')}`);
```

The `.update()` is unnecessary; commands automatically update rooms.

Comment (panpawn) [inline commands.js]: Done

**Comment by panpawn** [comment]
Updated

**Comment by Zarel** [comment]
I still think this is going to cause more problems than it solves, but I guess we can try it out and see.

**Pull request merged**